prow/crier/controller_test.go

```diff
@@ -0,0 +1,268 @@
+			jobsOnQueue: []string{"foo"},
+			knownJobs: map[string]*prowv1.ProwJob{
+				"foo": {
+					Spec: prowv1.ProwJobSpec{
+						Job: "foo",
+					},
+					Status: prowv1.ProwJobStatus{
+						State: prowv1.TriggeredState,
+					},
+				},
+			},
+			shouldReport: func(*prowv1.ProwJob) bool {
+				return true
+			},
+			expectReports: []string{"foo"},
+			expectPatch:   true,
+		},
+		{
+			name:        "doesn't report when it shouldn't",
+			jobsOnQueue: []string{"foo"},
+			knownJobs: map[string]*prowv1.ProwJob{
+				"foo": {
+					Spec: prowv1.ProwJobSpec{
+						Job: "foo",
+					},
+					Status: prowv1.ProwJobStatus{
+						State: prowv1.TriggeredState,
+					},
+				},
+			},
+			shouldReport: func(*prowv1.ProwJob) bool {
+				return false
+			},
+		},
+		{
+			name:        "doesn't report nonexistant job",
+			jobsOnQueue: []string{"foo"},
+			knownJobs:   map[string]*prowv1.ProwJob{},
+			shouldReport: func(*prowv1.ProwJob) bool {
+				return true
+			},
+		},
+		//{
+		//	name: "nil job panics",
+		//	jobsOnQueue: []string{"foo"},
+		//	knownJobs: map[string]*prowv1.ProwJob{
+		//		"foo" : nil,
+		//	},
+		//	shouldReport: func(*prowv1.ProwJob) bool {
+		//		return true
+		//	},
+		//},
+		{
+			name:        "doesn't report empty job",
+			jobsOnQueue: []string{"foo"},
+			knownJobs: map[string]*prowv1.ProwJob{
+				"foo": {},
+			},
+			shouldReport: func(*prowv1.ProwJob) bool {
+				return true
+			},
+		},
+		{
+			name:        "duplicate jobs report once",
+			jobsOnQueue: []string{"foo", "foo", "foo"},
+			knownJobs: map[string]*prowv1.ProwJob{
+				"foo": {
+					Spec: prowv1.ProwJobSpec{
+						Job: "foo",
+					},
+					Status: prowv1.ProwJobStatus{
+						State: prowv1.TriggeredState,
+					},
+				},
+			},
+			shouldReport: func(*prowv1.ProwJob) bool {
+				return true
+			},
+			expectReports: []string{"foo"},
+			expectPatch:   true,
+		},
+		{
+			name:        "previously-reported job isn't reported",
+			jobsOnQueue: []string{"foo"},
+			knownJobs: map[string]*prowv1.ProwJob{
+				"foo": {
+					Spec: prowv1.ProwJobSpec{
+						Job: "foo",
+					},
+					Status: prowv1.ProwJobStatus{
+						State: prowv1.TriggeredState,
+						PrevReportStates: map[string]prowv1.ProwJobState{
+							reporterName: prowv1.TriggeredState,
+						},
+					},
+				},
+			},
+			shouldReport: func(*prowv1.ProwJob) bool {
+				return true
+			},
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			q := kube.RateLimiter(controllerName)
+			for _, job := range test.jobsOnQueue {
+				q.Add(job)
+			}
+
+			inf := fakeInformer{
+				jobs: test.knownJobs,
+			}
+
+			rp := fakeReporter{
+				shouldReportFunc: test.shouldReport,
+			}
+			cs := fake.NewSimpleClientset()
+			nmwrk := 2
+			wg := &sync.WaitGroup{}
+			c := NewController(cs, q, inf, &rp, nmwrk, wg)
+
+			stopCh := make(chan struct{})
+			go c.Run(stopCh)
+
+			wait.Poll(10*time.Millisecond, testTimeout, func() (done bool, err error) {
```

Comment:
:-/

Reply:
The crier controller is not built to tell you when it's done processing things. In its view, it's never done; new things could be added to the queue at _any time_. This seemed like the best solution.